### ISSUE TITLE
Typography control refactor

### ISSUE THREAD

**Issue by pixedelic** [themeum/kirki#686]
Hello,

I'm trying to use this workaround https://jsfiddle.net/pixedelic/b3xro8rs/2/ to make data attributes available for options with selectize plugin. On JSfiddle works fine, but I can't make it work with Kirki... did you hack selectize somehow? Any filter that could create a conflict?

Thank you :-)

**Comment by aristath** [comment]
> I'm trying to use this workaround https://jsfiddle.net/pixedelic/b3xro8rs/2/ to make data attributes available for options with selectize plugin

Looks promising!

> did you hack selectize somehow? Any filter that could create a conflict?

Not that I can think of...
I'm assuming you're using this in https://github.com/aristath/kirki/blob/master/assets/js/controls/typography.js and then re-compiling the JS via grunt?

**Comment by pixedelic** [comment]
No, I created a new control type to manage at once font families and their relative variants/subsets.

However please, close this ticket. I'll try a new approach without using data attributes, so I guess that to make further investigations would be a waste of time.

Thanks

**Comment by aristath** [comment]
I was thinking that perhaps we can use pass the fonts as an array in our script, including all its arguments (like the email contacts example on the http://selectize.github.io/selectize.js/ )
This way we could render the select however we wanted, and we'd be able to improve things a lot

**Comment by aristath** [comment]
@pixedelic if you checkout the develop branch you'll see what I want to do...
I started re-implementing the typography control.
So now the font-families are not added in the control itself but rather populated by JS.
I'm exposing all font-properties in JS (variants, subsets etc) so we could work together on this one and achieve what we both want.

**Comment by aristath** [comment]
ok, another update:
I managed to figure out the basics:
When the user selects a font-family I'm getting the font-properties [here](https://github.com/aristath/kirki/blob/f89c22cfadc762b1b3b4f052d83444503bb2d855/assets/js/controls/typography.js#L75-L81).
Now all that remains is adding a subsets control there, and figuring out how to refresh the values of these controls with the ones available in this font.

**Comment by pixedelic** [comment]
Checking at this very moment. Looks great. I'm still trying to understand some things (I don't have your skills, sorry, and in particular I'm not familiar with GIT). However I have a completely different approach to fonts: Google (or other font providers like Typekit) doesn't work in terms of font weights, but in terms of font variants. Maybe this could seem a stupid difference, but for many fonts there's no italic style, so this could be a scenario:

you select "Droid Sans" for the body font, but the first time you have to use italic across the site, it will be the browser that will render it oblique. I'm a designer more than a developer and I hate when it happens.

And this is another very common scenario for me: if you have to decide the font of the main navigation menu, maybe you need one weight only, but very often, for the body text, I prefer to use a weight of 300 as regular weight and 700 or 800 as bold weight. With the previous version of Kirki I created a custom field that worked in this way.

I'd be glad to help you with this great project, you made a brilliant work, but I'm afraid I don't have the skills you need and, in particular, I don't think we are going two different ways with the font selector, not because you're wrong, of course, I think you are a WP hero, but because I need a completely different thing.

Maybe a donation will be more useful than my help :-)

**Comment by aristath** [comment]
Well, there's more than one ways to contribute... donations and code are just 2 of many.
In this case what would be more useful is if you could just describe the **ideal** typography control you have in your mind.
The only use of the typography control is in themes. So that's where we have to focus on this one.
Theme users want to design their site with ease, and most of them don't care about whether the font-weight is 300 or 700, they want to see "Light", "Book", "Ultra Bold Italic" etc.
A coder wouldn't bother with those things, they'd simply build a custom theme for their site. What a coder will do however, is build a theme for designers and end-users.

No matter how how look at it, the consumer of this control is designers & simple users. 
I'm neither, so your approach may be (and probably is) better.

As for my "skills", I'm just now learning JS... I'm good with PHP but I have to spend 30 minutes googling before I write a single line of JS, so don't underestimate yourself. :smile: 

EDIT: and don't worry about backwards-compatibility with users already using the old format of the control, I can take care of that internally.

**Comment by pixedelic** [comment]
Ha-ha, in this case I'll try to start from your approach to achieve what I have in mind and, if I get a decent result, I'll show you. Thank you for your time

**Comment by pixedelic** [comment]
However selectize has got a lot of limitations: no disabled attributes, no data attributes, it is very hard to work with it (and it also seems poorly supported at this moment)...

**Comment by aristath** [comment]
the problem with select2 and chosen is that a lot of other plugins also use them and they either use old & deprecated versions of those scripts, or they simply set options that caused anything else using them to conflict.
Switching to selectize was not a decision taken lightly. It was the best option available to resolve plugin conflicts, it's well-written and minimal, easy to style, and decently maintained.

btw I managed to get the font-weights to refresh properly.
Now on initial page load it looks for the available font options.
If no valid font-weight is saved in the db, then it uses "regular".
If "regular" does not exist in the font's variants then it uses the 1st one available.
Same thing happens when switching to a new font-family. It looks for the current value in the new family's variants, if not found falls back to regular, and if regular does not exist then uses the 1st available option.
The issue I have now is that changing the font-weight is not detected by the customizer as a change.
Once this is solved too, I think these are the changes that I'm going to make, let me know what you think:
- combine font-weight & font-style in a "variant" dropdown (Light, Regular, Regular Italic, Bold, etc)
- remove the units-selection dropdown from letter-spacing. Convert to a simple number field since letter-spacing is always in pixels
- combine the 2 fields of the font-size control to a single text control. Values will be entered including units and there will simply be a very efficient sanitization_callback in effect that will make sure a valid CSS unit is always used.

Thoughts?

**Comment by pixedelic** [comment]
Great!

> If no valid font-weight is saved in the db, then it uses "regular".

Does it means you allow to select among all the font variants, even if they don't exist for that particular font? I mean, are you working on [kirki/assets/js/controls/typography.js](url) or on [kirki/assets/js/customizer.js](url) too? Because I think you should remove the ability to select the unavailable font variants directly from the customizer page and select there the first variant available.

About changes: 

> combine font-weight & font-style in a "variant" dropdown (Light, Regular, Regular Italic, Bold, etc)

Wow, it should be great, even if using words can creates confusion: http://www.webtype.com/info/articles/fonts-weights/

> remove the units-selection dropdown from letter-spacing. Convert to a simple number field since letter-spacing is always in pixels

Unfortunately letter spacing also supports relative units and I often use those https://css-tricks.com/almanac/properties/l/letter-spacing/

> combine the 2 fields of the font-size control to a single text control. Values will be entered including units and there will simply be a very efficient sanitization_callback in effect that will make sure a valid CSS unit is always used.

Yes. For the use I make of the typography in my themes, I think you should provide the ability to decide to the theme developer, when it adds the field, if the font size will be intended in pixels, rem or em etc. and not to the user as you do now.

**Comment by pixedelic** [comment]
Thank you for this, of course, as always!

**Comment by aristath** [comment]
> I think you should remove the ability to select the unavailable font variants directly from the customizer page and select there the first variant available.

Oh I removed that... But imagine this scenario:
User has "Roboto" activated, with a font-weight of 100.
Then, they switch to "Lobster". However Lobster does not have a variant == 100, So we can either show it blank and pray they see it, or automatically make the switch to the best available option (regular-400/1st available option)

**Comment by aristath** [comment]
> Unfortunately letter spacing also supports other relative units and I often those https://css-tricks.com/almanac/properties/l/letter-spacing/

Yeah, I know it supports it... but I've never seen anyone use them for letter spacing. I use em & rem everywhere and hate pixels in general, but have you ever used a letter-spacing of 0.002em for example?

In any case, I see your point. I'll probably make that a textfield too and use the same sanitization I had in mind for the font-size

**Comment by pixedelic** [comment]
Yes, site title here, my last theme, for instance: http://www.pixedelic.com/themes/logan/ is 0.15em As I said I use it often. So: thank you, again.

**Comment by aristath** [comment]
@pixedelic I've been working on this non-stop for 9 hours and I think I've got it working (rewrote the JS from scratch).
it's not perfect yet, but it's getting there.
Now the biggest issue is the visual part: How on earth am I going to make this pretty? 
Any ideas on how to make this a bit more condensed?
font-family, variants & subsets all need their separate lines.
![customize wordpress dev site just another wordpress site](https://cloud.githubusercontent.com/assets/588688/13376258/1ad11f1a-ddc0-11e5-928a-56066da22cdd.jpeg)

**Comment by aristath** [comment]
LMAO it was easier than I expected:
![customize wordpress dev site just another wordpress site](https://cloud.githubusercontent.com/assets/588688/13376326/0bddb886-ddc2-11e5-981f-1391d6b9abd0.jpeg)

**Comment by pixedelic** [comment]
Yes exactly, almost the same idea, but yours is much better, easier, cleaner and more performing I guess. You're star. However, here it is:

![kirki-font-ui](https://cloud.githubusercontent.com/assets/527057/13376640/773bca88-ddc1-11e5-9c73-df17cfb804e1.png)

**Comment by aristath** [comment]
Since we're completely rebuilding this control, any ideas on how to further improve it?
Is it close to what you had in mind?

**Comment by ghost** [comment]
Aris and Pix... thank you both very much for dedicating so much time into
the typography control. Looks fantastic.

**Comment by pixedelic** [comment]
@aristath it seems fantastic to me, I don't think you should spend more time on this in a toolkit, other functions should come from theme authors.
Really great work! I'll play a little with it and, if you don't mind, I could suggest maybe some hooks.

Thank you so much.

And @cmwwebfx I haven't done anything at all :-)

**Comment by aristath** [comment]
@pixedelic @cmwwebfx Can you guys please download a fresh copy of the develop branch and test this out?
Let me know of any bugs you find :+1:

**Comment by pixedelic** [comment]
Sure, I'll let you know something ASAP

**Comment by pixedelic** [comment]
Maybe this thread should change title to allow other people to join it.

However, is this way to declare the field still valid (except for "units" that I removed)?

``` php
    Kirki::add_field( 'kirki_demo', array(
        'type'        => 'typography',
        'settings'    => 'typography_demo',
        'label'       => esc_attr__( 'Typography Control', 'kirki' ),
        'description' => esc_attr__( 'This is a composite typography control. It saves the data as an array, and you can define which of the controls you want shown.', 'kirki' ),
        'help'        => esc_attr__( 'This is a tooltip', 'kirki-demo' ),
        'section'     => 'my_section',
        'default'     => array(
            'font-style'     => array( 'bold', 'italic' ),
            'font-family'    => 'Roboto',
            'font-size'      => '14',
            'font-weight'    => '400',
            'line-height'    => '1.5',
            'letter-spacing' => '0',
            'color'          => '#333333',
        ),
        'priority'    => 10,
        'choices'     => array(
            'font-style'     => true,
            'font-family'    => true,
            'font-size'      => true,
            'font-weight'    => true,
            'line-height'    => true,
            'letter-spacing' => true,
            'color'          => true,
        ),
        'output' => array(
            array(
                'element' => 'body',
            ),
        ),
    ) );
```

Font weight selectors work great, they are removed and added properly on the UI, even if I don't see any change in the preview or in the front end, but maybe is a "to-do" thing.

My concern is with "size" "height" and "spacing"... first: labels are not correct. You did to gain space, of course, but maybe it should be better to split them in this way and use the correct form (it would create less problem with translations as well):

![kirki-ui-2](https://cloud.githubusercontent.com/assets/527057/13382964/016c0f96-de81-11e5-9717-9c63ac7e8433.png)

And about units: the user in this way is free to adopt pixels or ems or percentage etc., am I right? If they type numbers only?

**Comment by pixedelic** [comment]
And line height can also be a simple number, without unit, that it isn't the same of ems (http://stackoverflow.com/a/15135142)

**Comment by aristath** [comment]
> However, is this way to declare the field still valid (except for "units" that I removed)?

variants & subsets are now tied into the font-family. So if you show font-family, these will also be shown.
So this is what it would look like now:

``` php
    Kirki::add_field( 'kirki_demo', array(
        'type'        => 'typography',
        'settings'    => 'typography_demo',
        'label'       => esc_attr__( 'Typography Control', 'kirki' ),
        'description' => esc_attr__( 'This is a composite typography control. It saves the data as an array, and you can define which of the controls you want shown.', 'kirki' ),
        'help'        => esc_attr__( 'This is a tooltip', 'kirki-demo' ),
        'section'     => 'composite',
        'default'     => array(
            'font-family'    => 'Roboto',
            'font-size'      => '14px',
            'variant'        => '300italic',
            'line-height'    => '1.5',
            'letter-spacing' => '0',
            'color'          => '#333333',
        ),
        'priority'    => 10,
        'choices'     => array(
            'font-family'    => true,
            'font-size'      => true,
            'line-height'    => true,
            'letter-spacing' => true,
            'color'          => true,
        ),
        'output' => array(
            array(
                'element' => 'body',
            ),
        ),
    ) );
```

Though I am thinking of perhaps removing completely the `choices` argument... If you want to show a control, you just define a default value.
If no default values are defined, then all are shown. It should simplify things a lot.

> Font weight selectors work great, they are removed and added properly on the UI, even if I don't see any change in the preview or in the front end, but maybe is a "to-do" thing.

Yup, I just finished the 3rd rebuilt of the google-fonts management class, it works fine now. :+1: 

> My concern is with "size" "height" and "spacing"... first: labels are not correct. You did to gain space, of course, but maybe it should be better to split them in this way and use the correct form (it would create less problem with translations as well):

Done.
The only thing that may seem a bit "weird" is that when you click on the color selector, it pops down so that it fits. Well, we can't have it all...

> And about units: the user in this way is free to adopt pixels or ems or percentage etc., am I right? If they type numbers only?

They can't type numbers only... font-size for example can't be a simple number. So try this to see the sanitization function in action:
Enter font-size: 14, save and then refresh. You should see that 14 becomes 14px, so it forces you to enter a unit.
Similarly if you enter 2ems it will become 2em. Sanitization is awesome!!

> And line height can also be a simple number, without unit, that it isn't the same of ems (http://stackoverflow.com/a/15135142)

I know... line height is the only one there that doesn't enforce the use of units. So if you enter 2, it will remain 2. :+1: 

Try testing it again... everything should be working now.

**Comment by pixedelic** [comment]
> Though I am thinking of perhaps removing completely the choices argument... If you want to show a control, you just define a default value.
> If no default values are defined, then all are shown. It should simplify things a lot.

Seems great

> Similarly if you enter 2ems it will become 2em. Sanitization is awesome!!

Yes it is :-) However maybe it should be to replicate the same thing via javascript to validate the original value as well, I mean when you blur the field... just a suggestion, as always, don't kill me

> The only thing that may seem a bit "weird" is that when you click on the color selector, it pops down so that it fits. Well, we can't have it all...

And what if you added another field that I use often?

![kirki-ui-3](https://cloud.githubusercontent.com/assets/527057/13389206/3f73174a-dec6-11e5-96d7-26319bd90010.png)

**Comment by ghost** [comment]
Again, thank you guys for the huge effort you have put into rewriting the
font tool. I am away in India for 2 weeks for wedding I will test it out as
soon as I get a chance.

_Regards,_
_Ciaran Whelan_

Mobile: 0458 811 910
Skype: cmwwebfx
Twitter: cmwwebfx

**Comment by aristath** [comment]
I think we can close this one now...

The only big issue remaining is https://github.com/aristath/kirki/issues/717
which I've tried - without any success - to resolve for a number of days, so now it's on bountysource and hopefully some JS guru will be able to help us.

**Comment by pixedelic** [comment]
I'm not at home until monday, but I'll take a look both on this (did you consider the idea of adding uppercase in font module?) and on #717 as well. Thank you for your time, as always

**Comment by aristath** [comment]
> did you consider the idea of adding uppercase in font module?

Yup, I did and rejected the idea.
Adding upper/lowercase is as simple as it can get... No need to add all the available typography-related options in a single control. I wanted this to be minimal and functional.
If needed, uppercase/lowercase/normal is a simple radio-buttonset control that can be placed below a typography control. :+1:

**Comment by pixedelic** [comment]
Ouch... yes, but it would fix the bad look of the color palette that goes below when you click on it... no way to add it with a hook I guess?  Ok, shame, I'll survive :-)

**Comment by aristath** [comment]
> Ouch...

Sorry... I didn't mean to be rude in the above message, got a migraine and it can have that effect.

> no way to add it with a hook I guess?

Well, we could add a hook to allow adding the markup for that field, but it would also need some JS to work, which can't be added via a hook in the control... Which is why I don't have any hooks to add things in the controls themselves.
If you have any ideas to make this easier I'm all ears!

**Comment by enkota** [comment]
Just got round to trying this now. Very cool! Do you think it would make sense for the text boxes to auto update when a user changes the size? Currently you have to deselect the box for it to update.

Also the font weight you currently have `300italic` when I tried defaulting to `400` or `400normal` on `'Source Code Pro` it doesn't appear to work. Am I doing it wrong?

Is it possible to hide the subsets too?

**Comment by aristath** [comment]
> Do you think it would make sense for the text boxes to auto update when a user changes the size? Currently you have to deselect the box for it to update.

Done.

> Also the font weight you currently have 300italic when I tried defaulting to 400 or 400normal on 'Source Code Pro it doesn't appear to work. Am I doing it wrong?

Perhaps this is related to https://github.com/aristath/kirki/issues/717 ?

> Is it possible to hide the subsets too?

No. If a google font has subsets they will be displayed. This way sites that don't use english as their primary language will be able to select their language.

What I did though is this:
- removed latin from the list of subsets. Google always includes it anyway, no matter what the user chooses
- If only 1 variant is available, I'm hiding the variants select
- Since I removed latin from the subsets, the remaining languages are always non-latin, so if the font has a subset it will be displayed.
